SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include <inttypes.h>
+
+#include "Interop/Connection/SpatialTraceEventBuilder.h"
+#include "SpatialGDKSettings.h"
+#include <WorkerSDK/improbable/c_io.h>
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	SpatialEventTracer* EventTracer = static_cast<SpatialEventTracer*>(UserData);
+
+	Io_Stream* Stream = EventTracer->Stream.Get();
+	if (!ensure(Stream != nullptr))
+	{
+		return;
+	}
+
+	uint32_t ItemSize = Trace_GetSerializedItemSize(Item);
+	if (EventTracer->BytesWrittenToStream + ItemSize <= EventTracer->MaxFileSize)
+	{
+		EventTracer->BytesWrittenToStream += ItemSize;
+		int Code = Trace_SerializeItemToStream(Stream, Item, ItemSize);
+		if (Code != 1)
+		{
+			UE_LOG(LogSpatialEventTracer, Error, TEXT("Failed to serialize to with error code %d (%s"), Code, Trace_GetLastError());
+		}
+	}
+	else
+	{
+		EventTracer->BytesWrittenToStream = EventTracer->MaxFileSize;
+	}
+}
+
+SpatialScopedActiveSpanId::SpatialScopedActiveSpanId(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId)
+	: CurrentSpanId(InCurrentSpanId)
+	, EventTracer(InEventTracer->GetWorkerEventTracer())
+{
+	if (InCurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetValue());
+	}
+}
+
+SpatialScopedActiveSpanId::~SpatialScopedActiveSpanId()
+{
+	if (CurrentSpanId.IsSet())
+	{
+		Trace_EventTracer_ClearActiveSpanId(EventTracer);
+	}
+}
+
+SpatialEventTracer::SpatialEventTracer(const FString& WorkerId)
+{
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		if (Settings->bEventTracingEnabled)
+		{
+			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+			Enable(WorkerId);
+		}
+	}
+}
+
+SpatialEventTracer::~SpatialEventTracer()
+{
+	if (IsEnabled())
+	{
+		Disable();
+	}
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent)
+{
+	return TraceEvent(SpatialTraceEvent, nullptr, 0);
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId Causes)
+{
+	return TraceEvent(SpatialTraceEvent, &Causes, 1);
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const TArray<worker::c::Trace_SpanId>& Causes)
+{
+	return TraceEvent(SpatialTraceEvent, Causes.GetData(), Causes.Num());
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(FSpatialTraceEvent SpatialTraceEvent, const worker::c::Trace_SpanId* Causes,
+													   int32 NumCauses)
+{
+	if (!IsEnabled())
+	{
+		return {};
+	}
+
+	Trace_SpanId CurrentSpanId;
+	if (NumCauses > 0)
+	{
+		CurrentSpanId = Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses);
```

Comment:
Do these need to be called in the same place as Add_Event is called. If not then we might be able to get rid of the overloads by calling AddSpan whenever we have two spans and only want one. Then we just pass in the span ID associated with the event in the arguments to this function. We could then even put the event as a field in the event as the worker sdk does.

Reply:
I think we could potentially split up AddSpan and AddEvent. What would be the advantage of this approach?

Reply:
An event is a set of strings and a single span.
In the few places where you need to merge two span IDs you call some function EventTracer::MergeSpans that returns a new span ID.
If we have lots of cases with more than two then that can over an array overload.

But the main thing this would change would be that in combination with removing field IDs we could simplify the update stuff a lot.

Reply:
Or put another way. You don't ever really want to be holding onto lots of spans. There might be some places where that is how things fall out anyway and in that case fine. But generally you want to hold onto as few bits of information as possible. And so you want to turn two spans into one as soon as you can.
This doesn't work if we want to associate all of those with a custom message. So in that case then we'd have to keep this but it doesn't seem like that's what we are doing.